---
title: "Doing Parentage and finding Full-Sibs on the Kelp Rockfish"
output: 
  html_notebook:
    toc: true
---


This is a first stab at parentage on kelp rockfish.  In a previous document, `./Rmd/01-aggregate-genos-and-gsi-to-species.Rmd` 
I aggregated genotypes and filtered high missers and did GSI to identify kelp-rockfish amongst the juveniles
and the adults. I saved those results and use them here...


To start off with, let's load data and libs:
```{r load-stuff}
library(tidyverse)
library(CKMRsim)
library(stringr)


meta <- readRDS("../data/processed/meta-data-tibble.rds")
genos <- readRDS("../extdata/processed/genos-aggregated-and-no-hi-missers.rds")
juvie_species <- readRDS("../data/processed/juvie-gsi-sim-assignments.rds")   # gsi assignments of juvenile fish
adult_species <- readRDS("../data/processed/adult-gsi-sim-self-assignments.rds")   # gsi self-assignments of adult fish
```


## Selecting our fish and designating allele nomenclature

We gotta pick out the _S. atrovirens_.  For adults that is mostly done, but there are some
obvious misassigmnments, so we use the GSI results for both adults and juvies.  
```{r define-kelps}
kelp_ad <- adult_species %>% 
  filter(species == "atrovirens",
         score > 99) %>%
  select(NMFS_DNA_ID) %>% unlist() %>% unname()

kelp_juv <- juvie_species %>%
  filter(species == "atrovirens",
         score > 99) %>%
  select(NMFS_DNA_ID) %>% unlist() %>% unname

```

That is `r length(kelp_ad)` adults and `r length(kelp_juv)` juveniles.  

With those we can just filter down the genos to the ones that we want, and then we
can get it into the format required for CKMR.
```{r kelp-genos}
kelp_genos <- genos %>%
  filter(NMFS_DNA_ID %in% c(kelp_ad, kelp_juv))

# we will use this some more
kg2 <- kelp_genos %>% 
  select(NMFS_DNA_ID, locus, allele) %>%
  mutate(Chrom = "GTseq") %>% 
  mutate(Pos = as.integer(factor(locus, levels = unique(locus)))) %>%
  rename(Locus = locus,
         Allele = allele) %>%
  select(NMFS_DNA_ID, Chrom, Locus, Pos, Allele) %>%
  ungroup()

# get the allele freqs
kg_ckmr_markers <- kg2 %>%
  filter(!is.na(Allele)) %>% # it is vital to filter out the NAs at this stage
  group_by(Chrom, Locus, Pos, Allele) %>%
  summarise(counts = n()) %>%
  group_by(Locus, Pos) %>%
  mutate(Freq = counts / sum(counts)) %>%
  select(-counts) %>%
  mutate(AlleIdx = 1,
         LocIdx = 1) %>%
  reindex_markers(.)
```
Great! We have a total `r nrow(kg_ckmr_markers)` alleles in this data set.


## Doing CKMR simulations to assess power

Really simply let's just do this to see what our power for parentage ought to look like.
```{r CKMR-create}
CK <- create_ckmr(kg_ckmr_markers, kappa_matrix = kappas[c("PO", "FS", "HS", "U"), ])

# see what it tells us when we print it:
CK
```

Then do the simulations to see what the false positive rates are for finding parents when the truth
is U or FS:
```{r CKMR-sim-and-sample}
Qs <- simulate_Qij(C = CK, froms = c("PO", "FS", "HS", "U"), tos = c("PO", "FS", "HS", "U"), reps = 10^4)

# then do the importance sampling to get the FPRs
mc_sample_simple(Qs, nu = "PO", de = c("U", "FS"), tr = c("U", "FS"), method = "both")
```
We can come back to that later to see about finding Full Sibs.

## Making genotype matrices
```{r}
kelp_haps <- kelp_genos %>%
  filter(!is.na(allele)) %>%  # once again, it is critical to remove these at this point
  select(NMFS_DNA_ID, locus, gene_copy, allele) %>%
  rename(Locus = locus, Allele = allele)

kelp_idx_frame <- kg_ckmr_markers %>%
  select(Locus, Allele, LocIdx, AlleIdx) %>%
  group_by(Locus) %>%
  mutate(NumA = n()) %>%  # get the number of alleles at each locus
  ungroup() %>%
  left_join(kelp_haps, .)  %>%  # join the alle_idx's onto the actual genotype data
  select(NMFS_DNA_ID, Locus, gene_copy, LocIdx, NumA, AlleIdx) %>%
  spread(key = gene_copy, value = AlleIdx) %>%
  mutate(GenoIdx = index_ab(a = `1`, b = `2`, A = NumA))

# make a matrix of genotype integers 
wide_kelp <- kelp_idx_frame %>%
  select(NMFS_DNA_ID, LocIdx, GenoIdx) %>%
  spread(data = ., key = LocIdx, value = GenoIdx)
```

Don't forget to set NA's to 0, and then decrease each value by 1:
```{r make-mat}
rocky_mat <- as.matrix(wide_kelp[, -1])
rownames(rocky_mat) <- wide_kelp$NMFS_DNA_ID
rocky_mat[is.na(rocky_mat)] <- 0
rocky_mat <- rocky_mat - 1
storage.mode(rocky_mat) <-  "integer"
```

## Looking for duplicated samples

We can quickly look through rocky_mat for pairs of indivs with lots of matching genotypes.
```{r check-for-dupes}
matchers <- pairwise_geno_id(S = rocky_mat, max_miss = 12) %>%
  arrange(num_mismatch) %>%
  mutate(NMFS_DNA_ID_1 = rownames(rocky_mat)[ind1],
         NMFS_DNA_ID_2 = rownames(rocky_mat)[ind2])
matchers
```
We need to deal with these somehow.  But before we do that, I want to use these numbers to come up with
a per-genotype discordance rate.  That should be pretty straightforward:
```{r compute-discordance-rate}
sum(matchers$num_mismatch) / sum(matchers$num_loc)

```
So, to repeat, the discordance rate across all these multiply-genotyped individuals is
`r sum(matchers$num_mismatch) / sum(matchers$num_loc)`---roughly 3 in 1000.  That is a nice
number.

Now, to deal with these duplicates, here is what I will do: I am going to say that anyone with 3 or fewer mismatches is certainly 
the same individual, and we are going to take just one from amongst those pairs/groups.  We use a hasty method:
we toss out all the individuals that appear in both ind1 and ind2, and then all the remaining ind2's.  This 
might throw out one more than we need to, but it is what we do for now.
```{r toss-matchers}
def_same <- matchers %>%
  filter(num_mismatch <= 3)

toss_these <- c(intersect(def_same$ind1, def_same$ind2), def_same$ind2) %>%
  unique()

rocky_mat_tossed <- rocky_mat[-toss_these,]
```

Now we have to break that into two matrices: one for the adults and one for the kids.
This should not be too hard.  We have to deal with the fact that not all of the adults in the meta data
had few enough missing genotypes to be included.
```{r break-stage}
kelp_ad_kept <- kelp_ad[kelp_ad %in% rownames(rocky_mat_tossed)]
kelp_juv_kept <- kelp_juv[kelp_juv %in% rownames(rocky_mat_tossed)]

adult_mat <- rocky_mat_tossed[kelp_ad_kept, ]
juvie_mat <- rocky_mat_tossed[kelp_juv_kept, ]
```


## Before doing the comparisons, compute internal heterozygosities

We want to have these to make sure we don't have contaminated individauls.
```{r compute-int-hzs}
int_hz <- kg2 %>% 
  group_by(NMFS_DNA_ID, Locus) %>% 
  summarise(hz = Allele[1]!=Allele[2]) %>% 
  filter(!is.na(hz)) %>% 
  group_by(NMFS_DNA_ID) %>% 
  summarise(num_loc = n(), num_hz = sum(hz), fract_hz = num_hz / num_loc)
```
And now let us plot that .
```{r plot-int-hz, fig.width=12}
ggplot(int_hz, aes(x = fract_hz)) +
  geom_density(fill = "orange", alpha = 0.3)
```

So, we would be suspicious of individuals with int_hz > .6 or less than 0.28.

## Doing the comparisons

We need to flatten the CKMR object appropriately
```{r flatten}
# flatten those out
po_flat <- flatten_ckmr(CK, "PO")
unrel_flat <- flatten_ckmr(CK, "U")

# then compute the log-likelihoods for the parent offspring relationship
po_logl_flat <- po_flat
po_logl_flat$probs <- log(po_flat$probs / unrel_flat$probs)

```


Now we are ready to do these!
```{r do-comps}
idx <- 1:nrow(juvie_mat)
names(idx) <- idx
comps <- lapply(idx, function(i) {
    tmp <- comp_ind_pairwise(S = adult_mat, T = juvie_mat, t = i, values = po_logl_flat$probs, nGenos = po_logl_flat$nGenos, Starts = po_logl_flat$base0_locus_starts)
    #tmp[rev(top_index(tmp$value, 5)), ]  # just take the top 5 from each
    }) %>%
  dplyr::bind_rows(.id = "offspring") %>%
  dplyr::tbl_df() 

# make int_hz something we can join on there
tmp1 <- int_hz %>% rename(id1 = NMFS_DNA_ID, fract_hz1 = fract_hz) %>% select(id1, fract_hz1)
tmp2 <- int_hz %>% rename(id2 = NMFS_DNA_ID, fract_hz2 = fract_hz) %>% select(id2, fract_hz2)

# have a look at that
result <- arrange(comps, desc(value)) %>%
  mutate(number = 1:nrow(.)) %>%
  mutate(id1 = rownames(juvie_mat)[as.integer(offspring)],
         id2 = rownames(adult_mat)[ind]) %>%
  left_join(tmp1) %>%
  left_join(tmp2)

result %>% filter(number <= 1000)

```
It is clear that the internal heterozygosities are reasonable, so it is not like any of these likely 
pairs are from contaminated individuals.

Try plotting all those pairs, and compare them to the simulated values
```{r plot-all-pairs-density, fig.width = 6}
simlogls <- extract_logls(Q = Qs, numer = c(PO = 1), denom = c(U = 1))
ggplot(result, aes(x = value)) +
  geom_density(colour = "red") +
  geom_density(data = simlogls, aes(x = logl_ratio, fill = true_relat), alpha = 0.25) +
  scale_x_continuous(breaks = seq(-170, 60, by = 10))
```
OK, that is nice to see.  It is a little hard to make out---the red line above is the
observed distribution from all the pairs.  You see that it lands almost perfectly on
top of the simulated distribution for unrelated pairs.  That is good! It looks super clean.  





## Checking the logls

As long as we are at it, we may as well plot the distribution of our LogL>10 individuals 
against the backdrop of our simulated PO/U LogL values for true parent offspring pairs, to make sure that
things look reasonable.
```{r plot-po-logls}
good_uns <- result %>% 
  filter(value > 10) %>%
  mutate(y = 0.001)
po_frame <- simlogls %>% 
  filter(true_relat == "PO")
ggplot() +
  geom_density(data = po_frame, aes(x = logl_ratio), fill = "green", alpha = 0.3) +
  geom_point(data = good_uns, aes(x = value, y = y))

```

Note that the two with the lowest LogLs are kinda lower than we would really expect, but
not totally impossible.  Check which ones those are again:
```{r print-good-uns}
good_uns
```
But, otherwise, those values look pretty reasonable.  

## Could they be other types of relatives?

We have seen above that most full siblings  can have PO/U logls that are right around 10.  Half-siblings can also
see PO/U logls in that ballpark.  It is super unlikely that you would have an adult full sibling of anyone recruiting
this year, but it is possible that you could have a half-sibling.  I should look at the PO/HS and PO/FS LogLs for all
8 of our candidate PO pairs.  First off, let's see what those ratios look like for 
PO vs HS or FS.  
```{r plot-fs-ratios, fig.width = 8}
po_fs <- extract_logls(Qs, numer = c(PO = 1), denom = c(FS = 1)) 
ggplot(po_fs, aes(x = logl_ratio, fill = true_relat)) +
  geom_density(alpha = 0.25) +
  ggtitle("PO/FS Log-likelihood Ratios")
```

```{r plot-hs-ratios, fig.width = 8}
po_hs <- extract_logls(Qs, numer = c(PO = 1), denom = c(HS = 1)) 
ggplot(po_hs, aes(x = logl_ratio, fill = true_relat)) +
  geom_density(alpha = 0.25) +
  ggtitle("PO/HS Log-likelihood Ratios")
```

OK, so, not super easy to resolve these things, but it might be worth computing them.

### Computing PO/FS and PO/HS for the good_uns

First we need to filter down our genotype matrices to include only the offspring and the adults that
are in good_uns.
```{r filter-matrices}
adult_mat_filt <- adult_mat[good_uns$id2, ]
juvie_mat_filt <- juvie_mat[good_uns$id1, ]
```
Then flatten some more CKMR objects and compute the ratios:
```{r more-flattening}
fs_flat <- flatten_ckmr(CK, "FS")
hs_flat <- flatten_ckmr(CK, "HS")

po_fs_logl_flat <- po_flat
po_fs_logl_flat$probs <- log(po_flat$probs / fs_flat$probs)

po_hs_logl_flat <- po_flat
po_hs_logl_flat$probs <- log(po_flat$probs / hs_flat$probs)
```

Then do the comparisons.  This is fast because it is only a few of them.
```{r do-handful-of-comps-fs}
idx <- 1:nrow(juvie_mat_filt)
names(idx) <- idx
result_po_fs <- lapply(idx, function(i) {
    tmp <- comp_ind_pairwise(S = adult_mat_filt, T = juvie_mat_filt, t = i, values = po_fs_logl_flat$probs, nGenos = po_fs_logl_flat$nGenos, Starts = po_fs_logl_flat$base0_locus_starts)
    #tmp[rev(top_index(tmp$value, 5)), ]  # just take the top 5 from each
    }) %>%
  dplyr::bind_rows(.id = "offspring") %>%
  dplyr::tbl_df()  %>%
  filter(offspring == ind) %>%  # just filter it down to the ones we want here...
  rename(po_fs_logl = value) %>%
  mutate(id1 = rownames(juvie_mat_filt)[as.integer(offspring)],
         id2 = rownames(adult_mat_filt)[ind])

result_po_fs
```

And the same for half-sibling ratios:
```{r do-handful-of-comps-hs}
idx <- 1:nrow(juvie_mat_filt)
names(idx) <- idx
result_po_hs <- lapply(idx, function(i) {
    tmp <- comp_ind_pairwise(S = adult_mat_filt, T = juvie_mat_filt, t = i, values = po_hs_logl_flat$probs, nGenos = po_hs_logl_flat$nGenos, Starts = po_hs_logl_flat$base0_locus_starts)
    #tmp[rev(top_index(tmp$value, 5)), ]  # just take the top 5 from each
    }) %>%
  dplyr::bind_rows(.id = "offspring") %>%
  dplyr::tbl_df()  %>%
  filter(offspring == ind) %>%  # just filter it down to the ones we want here...
  rename(po_hs_logl = value) %>%
  mutate(id1 = rownames(juvie_mat_filt)[as.integer(offspring)],
         id2 = rownames(adult_mat_filt)[ind])

result_po_hs
```
So, R013607 is not looking super strong as a PO pair.  Could be a half-sibling.  There are so few of them that it
seems like we might want to consider confirming them with a ddRAD run or something.

All in all though, I think we do have a handful of very strong PO pairs.

## Look at the meta data for those fish

Have a quick looksie:
```{r po-meta}
dir.create("csv_outputs")
meta_trim <- meta %>%
  select(NMFS_DNA_ID, SAMPLE_ID, LENGTH, REPORTED_LIFE_STAGE, COLLECTION_DATE, SITE, LATITUDE_M, LONGITUDE_M, SAMPLE_COMMENTS, LOCATION_COMMENTS_M)

good_pairs <- good_uns %>%
  select(-offspring, -ind, -y, -number) %>%
  rename(offspring = id1, 
         parent = id2,
         po_u_logl = value) %>%
  select(offspring, parent, everything())

meta_juv <- meta_trim
names(meta_juv) <- paste("juv", names(meta_juv), sep = "_")
meta_ad <- meta_trim
names(meta_ad) <- paste("ad", names(meta_ad), sep = "_")

pairs_with_meta <- left_join(good_pairs, meta_juv, by = c("offspring" = "juv_NMFS_DNA_ID")) %>%
  left_join(., meta_ad, by = c("parent" = "ad_NMFS_DNA_ID"))

write_csv(pairs_with_meta, "csv_outputs/pairs_with_meta.csv")

pairs_with_meta
```



## Let's look for full-siblings

### Power for full-sib detection

First, we want to see what the distributions of FS/U LogLs look like for
FS, U, and HS.  This is pretty straightforward to do with the ckmr object we made:
```{r plot-fs-logls, fig.width=8}
fs_u <- extract_logls(Qs, numer = c(FS = 1), denom = c(U = 1))
ggplot(fs_u, aes(x = logl_ratio, fill = true_relat)) +
  geom_density(alpha = 0.25) +
  ggtitle("FS/U Log-likelihood Ratios")
```

And we can compute the false positive rates as well:
```{r fs-fprs}
mc_sample_simple(Qs, nu = "FS", de = c("U"), tr = c("U", "FS", "HS", "PO"), method = "both")
```

### Doing the pairwise comparisons

I am just going to bung the parents and juveniles together into a single genotype matrix of all the kelps
and then I will use my existing function with the same matrix as source and target, then filter out half of 
them.
```{r bung-compare-filter}
all_kelp_pre <- rbind(adult_mat, juvie_mat)

# keep only those with < 0.61 internal HZ
hz_keepers <- int_hz$NMFS_DNA_ID[int_hz$fract_hz < 0.61]
hz_keepers <- intersect(hz_keepers, rownames(all_kelp_pre))
all_kelp_mat <- all_kelp_pre[hz_keepers, ]

# then make the flattened thing
fs_u_flat <- fs_flat
fs_u_flat$probs <- log(fs_flat$probs / unrel_flat$probs)

# then do the comparisons, only keep the top 20 for each one
idx <- 1:nrow(all_kelp_mat)
names(idx) <- idx
fs_comps <- lapply(idx, function(i) {
    tmp <- comp_ind_pairwise(S = all_kelp_mat, T = all_kelp_mat, t = i, values = fs_u_flat$probs, nGenos = fs_u_flat$nGenos, Starts = fs_u_flat$base0_locus_starts)
    tmp[rev(top_index(tmp$value, 20)), ]
    }) %>%
  dplyr::bind_rows(.id = "member1") %>%
  dplyr::tbl_df() %>%
  rename(member2 = ind) %>%
  mutate(member1 = as.integer(member1)) %>%
  filter(member1 < member2) %>%
  mutate(name1 = rownames(all_kelp_mat)[member1],
         name2 = rownames(all_kelp_mat)[member2]) %>%
  arrange(desc(value))

```

OK, that is pretty cool.  But we really need a little metadata on there to know who these individuals are.
```{r tidy-fs-comps-meta}
mini_meta <- meta %>%
  select(NMFS_DNA_ID, REPORTED_LIFE_STAGE, COLLECTION_DATE, LATITUDE_M, LONGITUDE_M)
mm1 <- mini_meta
names(mm1) <- c("name1", "stage1", "date1", "lat1", "long1")
mm2 <- mini_meta
names(mm2) <- c("name2", "stage2", "date2", "lat2", "long2")

fs_comps_meta <- fs_comps %>%
  left_join(., mm1) %>%
  left_join(., mm2)

# and, don't forget that we have to toss out the parent-offspring pairs 
# that we found
dumpers <- data_frame(name1 = c(good_uns$id1, good_uns$id2),
                      name2 = c(good_uns$id2, good_uns$id1))
  
high_likely_fs <- fs_comps_meta %>%
  filter(value > 13) %>%
  anti_join(dumpers) %>%
  arrange(member1, member2)
  
high_likely_fs


```

So, that looks to be some full-sib pairs.  Let's check how many individuals appear more than once in that list:
```{r more-than-oncelers}
table(c(high_likely_fs$name1, high_likely_fs$nam2))
```
Nope! None o them appear more than once.  So, we have only pairs of full siblings here.  No big clusters.



## Dump data to output files



So, what we do here is dump a bunch of stuff
```{r assemble-output-list}

duplicate_samples <- def_same %>% 
  tbl_df() %>% 
  mutate(id1 = rownames(rocky_mat)[ind1], id2 = rownames(rocky_mat)[ind2])


output_list <- list(
  kelp_idx_frame = kelp_idx_frame,  # these are the haplotypes of each individual in numeric-allelic form
  kg_ckmr_markers = kg_ckmr_markers,   # allele freqs in CKMR long format
  top_po_pairs = result %>% filter(number <= 11),
  duplicate_samples = duplicate_samples,
  int_hz = int_hz,
  high_likely_fs = high_likely_fs
)

dir.create("rds_outputs")
saveRDS(output_list, file = "rds_outputs/kr-parentage-output-list.rds", compress = "xz")

```


## Look further at the siblings

Let's plot the sib-logls against what we think they should be
```{r  plot-sib-simlogls, fig.width=6}
fs_u_simlogls <- extract_logls(Qs, numer = c(FS = 1), denom = c(U = 1))
tmp_frame <- high_likely_fs %>%
  mutate(y_line = 0.002)
ggplot() +
  geom_density(data = fs_u_simlogls %>% filter(true_relat == "FS"), mapping = aes(x = logl_ratio), fill = "green", alpha = 0.3) +
  geom_point(data = tmp_frame, aes(x = value, y = y_line))
```
I still need to look at this stuff a little more.  There appears to be one duplicate in there, or something.  Namely `R014981` and `R014982`
have a ridiculously high full sib likelihood.  And they are right next to each other in NMFS_ID.  So something is weird with them.

```{r look-at-weirdos}
kg2 %>% 
  filter(NMFS_DNA_ID %in% c("R014981", "R014982")) %>%
  left_join(kg_ckmr_markers) %>%
  arrange(LocIdx, AlleIdx)
```
You know, I bet that somehow the barcodes for both of them ended up R014982's well:
```{r weirdo-int-hz}
int_hz %>% filter(NMFS_DNA_ID %in% c("R014981", "R014982"))
```

I should look at the actual read depths on these.  to confirm, but certainly the high fract_hz suggests something is going on.

Let's do that. R014982 was s6851 on gtseq_run 32 (that info is in the processed genos data)
```{r close-look-for-barcode-mixup}
rd_look <- readRDS("../extdata/haplot-rds/gtseq32_96loci.rds") %>%
  filter(id == "s6851")

rd_look %>% arrange(locus, desc(depth))
```

At some of the loci it has high read depths for more than two alleles.  So, I'll bet that is what is going on there.

There is another pair with a full-sib logl that is way higher than we would have expected, too.  

Perhaps the LogL sims will shed more light on these.


## LogL Sims for Each Individual

Having found these apparent parent-offspring and full sibling pairs it is very, very 
important to make sure that they are not just aberrant individuals (like non-target species, etc.).

We can do this by using gsi_sim's logl-sims method.  We are going to throw all the kelp adults and juveniles
into a baseline file for kelps, and then to baseline logl-sims on them and get the distribution of z-scores 
for everyone.   So, we need to make that gsi_sim file that has everyone in it:
```{r make-everyone-baseline}
all_kelps_for_base <- kelp_idx_frame %>%
  mutate(genostr = paste(`1`, `2`)) %>%
  select(NMFS_DNA_ID, Locus, genostr) %>%
  tidyr::spread(data = ., key = Locus, value = genostr)

all_kelps_for_base[is.na(all_kelps_for_base)] <- "0 0"

cat(nrow(all_kelps_for_base), ncol(all_kelps_for_base) - 1, "\n", file = "../gsi_area/all_kelps_baseline.txt")
cat(names(all_kelps_for_base[-1]), sep = "\n", file = "../gsi_area/all_kelps_baseline.txt", append = TRUE)
cat("POP all_kelps\n", file = "../gsi_area/all_kelps_baseline.txt", append = TRUE)
write.table(all_kelps_for_base, row.names = F, col.names = F, quote = F, sep = "\t", file = "../gsi_area/all_kelps_baseline.txt", append = TRUE)
```

And then I do the logl-sims.  This is set to eval=FALSE 'cuz it takes a long time.
```{sh compute-base-logl-sims}
cd gsi_area
~/Documents/git-repos/gsi_sim/gsi_sim-Darwin -b all_kelps_baseline.txt --self-assign  --base-logl-sims 1000 0 > logl_sims_stdout.txt
```
Then we can read in those logl-sims...
```{r read-logl-sims}
logl_sims <- read.table("../gsi_area/baseline_logl_summary.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  tbl_df()
```
and see what they look like.  So, let's just plot the ObsLogLs first:
```{r plot-obs-logls, fig.width=5}
ggplot(logl_sims, aes(x = ObsLogL)) +
  geom_histogram(binwidth = 1, fill = "blue")
```

Clearly there are some individuals off in the tails there.  I suspect the one that is way too high
is probably a contaminated sample.  The ones that are way too low might be different species, I would think.
Let's look at the z-scores while we are at it:
```{r plot-z-scores}
ggplot(logl_sims, aes(x = zScore)) +
  geom_histogram(binwidth = .1, fill = "blue")
```


Once again, clearly there are some that are way too low.  So what we need to do now is 
check the ObsLogLs and the zScores for the pairs that we found.


### Checking pairs against ObsLogL and zScore

```{r put-z-scores-on-po-pairs}
logl_sim_tmp_off <- logl_sims %>% 
  select(-AssignedTo) %>%
  rename(offspring = FishId, obslogl_off = ObsLogL, zscore_off = zScore, fsg_off = FractionSimmedGreater, num_loci_off = NumLoci)
logl_sim_tmp_par <- logl_sims %>% 
  select(-AssignedTo) %>%
  rename(parent = FishId, obslogl_par = ObsLogL, zscore_par = zScore, fsg_par = FractionSimmedGreater, num_loci_par = NumLoci)
po_pairs_with_base_logls <- good_pairs %>% 
  left_join(logl_sim_tmp_off) %>%
  left_join(logl_sim_tmp_par)
po_pairs_with_base_logls
```
Those all look like reasonable values.  Now, lets look at the full-siblings
```{r z-scores-on-fs-pairs}
logl_sim_tmp_1 <- logl_sims %>% 
  select(-AssignedTo) %>%
  rename(name1 = FishId, obslogl1 = ObsLogL, zscore1 = zScore, fsg1 = FractionSimmedGreater, num_loci1 = NumLoci)
logl_sim_tmp_2 <- logl_sims %>% 
  select(-AssignedTo) %>%
  rename(name2 = FishId, obslogl2 = ObsLogL, zscore2 = zScore, fsg2 = FractionSimmedGreater, num_loci2 = NumLoci)
fs_pairs_with_base_logls <- high_likely_fs %>%
  left_join(logl_sim_tmp_1) %>%
  left_join(logl_sim_tmp_2)
fs_pairs_with_base_logls
```


## Gotta look at our sampling intensities

We need to have a look at when all these fish were sampled so that we can
see whether the preponderance of same-year individuals is more than what
we expect by chance.
```{r meta-for-all-kelps}
kept_kelp_meta <- meta %>%
  filter(NMFS_DNA_ID %in% c(rownames(adult_mat), rownames(juvie_mat)))
```

Now, plot the samples obtained against day:
```{r plot-sampling}
kept_kelp_days <- kept_kelp_meta %>%
  group_by(REPORTED_LIFE_STAGE, COLLECTION_DATE) %>%
  tally()

ggplot(kept_kelp_days, aes(x = COLLECTION_DATE, y = n)) +
  geom_point() +
  facet_wrap(~REPORTED_LIFE_STAGE, scales = "free_y")
       
```

OK, let's summarize that all the way down to years:
```{r plot-sampling-by-years}
kept_kelp_years <- kept_kelp_meta %>%
  mutate(year = lubridate::year(COLLECTION_DATE)) %>%
  group_by(REPORTED_LIFE_STAGE, year) %>%
  tally() %>% 
  ungroup() %>%
  filter(year >= 2013) %>%
  rename(life_stage = REPORTED_LIFE_STAGE)

ggplot(kept_kelp_years, aes(x = year, y = n, fill = life_stage)) +
  geom_bar(stat = "identity") +
  facet_wrap(~life_stage)
  
```


Look at that.  So, if our method was totally broken and bogus and we were just
getting a boatload of artifactual false positives, we would be expecting a lot
more sib-pairs inferred between years.  But we don't.  Here is what we find
amongst the juveniles:
```{r count-juvie-pairs-years}
high_likely_fs %>% 
  filter(stage1 == "JUVENILE" & stage2 == "JUVENILE") %>%
  mutate(year1 = lubridate::year(date1),
         year2 = lubridate::year(date2)) %>%
  group_by(year1, year2) %>% 
  tally()
```

That is really interesting.  If siblings occurred at random within years, then the fraction of them
we would expect from each year would depend on the number of pairs that we could form, and those
numbers look like:
```{r expected-num-pairs}
kept_kelp_years %>%
  filter(life_stage == "JUVENILE") %>%
  mutate(num_pairs = n * (n - 1) / 2,
         fract_pairs = num_pairs / sum(num_pairs))
```
So, there are significantly fewer pairs than expected at random in 2015, or, I should say, there is definitely a different
distribution of sib-pairs in 2014 (lots) and 2015 (few).

While we are at it, we could compute the expected fraction of between-years assignments if these
things were all false positives (which they aren't, clearly). This is approximate, but pretty
darn close.
```{r expected-fractions}
tmp <- kept_kelp_years %>%
  filter(life_stage == "JUVENILE") 
freq <- tmp$n / sum(tmp$n)
names(freq) <- tmp$year
tmp2 <- outer(freq, freq) * 2
diag(tmp2) <- diag(tmp2) / 2
tmp2[upper.tri(tmp2)] <- 0

tmp2


```
So, the upshot there is that if our method was totally bogus and random and wrong,
we would expect to have most between-year assignments.  Which we clearly don't.

That is good!



