---
title: "Visually Checking Homozygosity"
output: 
  html_notebook:
    toc: true
---


## Introduction

I just want to make some pictures like what I did with some of the PIRE rad data
and the Lobster data that showed the remarkable heterozygote deficits.  

The hope is that I will be able to make the point that our microhaplotype data
do not suffer from the same issues.  And, at any rate, it is something that
should be investigated.


To get to this point we had to run through `./Rmd/01-aggregate-genos-and-gsi-to-species.Rmd` and also
`./Rmd/02-kelp-rockfish-parentage-and-sibs.Rmd`.  That last one creates an output file RDS that we can 
use here:
```{r load-data}
library(tidyverse)
library(CKMRsim)

dat <- readRDS("rds_outputs/kr-parentage-output-list.rds")
```


## Getting the data together

For each allele we have the allele frequencies. Let's name them something better, and while we
are at it, compute the expected frequency of homozygotes and heterozygotes for each 
allele.
```{r get-freqs}
freqs <- dat$kg_ckmr_markers %>%
  mutate(exp_homoz = Freq ^ 2,
         exp_het = 2 * Freq * (1 - Freq))
```

Now, we also want to put together the observed rates of homozygosity for each allele.
There will be some weird counting for the heterozygotes because some genotypes are placed into
two het categories, but that is OK. The expectations are still the same, even though things will 
be correlated.

First, get genotypes:
```{r get-genos}
genos <- dat$kelp_idx_frame %>%
  rename(alle1 = `1`,
         alle2 = `2`) %>%
  select(-GenoIdx, -NumA) %>%
  mutate(homoz = alle1 == alle2)

genos[1:20,]
```

Then compute the fraction of homozygous genotypes.  To do this we count up the number
of homozygous genotypes of each allelic type, and then divide by the total number
of genotypes at the locus. 
```{r count-homoz}
homoz_counts <- genos %>%
  group_by(LocIdx, alle1, homoz) %>%
  tally() %>%
  group_by(LocIdx) %>%
  mutate(tot_genos = sum(n),
         obs_homoz = n / tot_genos) %>%
  filter(homoz == TRUE) %>%
  rename(AlleIdx = alle1)
  
  
```

Then we need to join that to the freqs, 
and while doing that, it is going to be nice to compute the standard deviations around the
expected values, and a z-score, so we cqn make a little comparison of how far out the observed values fall.
```{r join-obs-exp}
exp_and_obs <- freqs %>%
  left_join(homoz_counts) %>%
  mutate(exp_SD = sqrt(exp_homoz * (1 - exp_homoz) / tot_genos),
         zscore = (obs_homoz - exp_homoz) / exp_SD)
```
```{r plot-z}
hist(exp_and_obs$zscore, breaks = 50)
```
It is interesting to look at those with Z-scores above 3.3, but we want to be sure
to filter so that we expected to see at least 50 of the homozygotes, 
so that we don't have a lot of the singletons in there.  (This is liable to miss a few
really low-frequency null alleles, but I am not that worried.)
```{r investigate-outliers}
nasties <- exp_and_obs %>% 
  filter(exp_homoz * tot_genos >= 50) %>%
  filter(zscore > 3.3) %>%
  select(Locus, LocIdx, AlleIdx, Freq, exp_homoz, obs_homoz, n, tot_genos, zscore) %>%
  mutate(quick_dirty = (obs_homoz - exp_homoz) / (2 * Freq))
nasties
```
Now, see which loci those were:
```{r nasty-loci}
nasty_loci <- unique(nasties$Locus)
nasty_loci
```

So, that is 10 loci.

While we are at it,  let's look at the total number of genotypes at each locus:
```{r count-tot-genos}
tot_by_locus <- exp_and_obs %>%
  group_by(Locus, LocIdx) %>%
  summarise(ngenos = mean(tot_genos, na.rm = TRUE))

tot_by_locus
```
And make a plot of that
```{r tot-hist}
hist(tot_by_locus$ngenos, breaks = 30)
```

Now, plot the homozygosities and color it according to whether there a likely nulls.  There are 10 loci we
identified as likely having null alleles.
```{r plot-it}
tmp <- exp_and_obs %>%
  mutate(likely_has_nulls = Locus %in% nasty_loci)

ggplot(tmp, aes(x = exp_homoz, y = obs_homoz, colour = likely_has_nulls)) +
  geom_abline(intercept = 0, slope = 1, linetype = "dotted") +
  geom_point()


fortalk <- ggplot(tmp, aes(x = exp_homoz, y = obs_homoz, colour = likely_has_nulls)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, linetype = "dotted") +
  xlab("Expected fraction of homozygotes") +
  ylab("Observed fraction of homozygotes") +
  guides(colour = FALSE)

ggsave(fortalk, filename = "microhap_homozyg.pdf", width = 4, height = 3)

```

That is pretty.  And let's have a look at it without any of the likely_has_nulls:
```{r plot-it2, fig.width = 6}
ggplot(tmp %>% filter(likely_has_nulls == FALSE), aes(x = exp_homoz, y = obs_homoz, colour = likely_has_nulls)) +
  geom_abline(intercept = 0, slope = 1, linetype = "dotted") +
  geom_point()

```

That is way better than RAD data!

And, finally, print out the data with the Z-scores (in other words observed homo freq - expected homo freq 
divided by the expected standard deviation)
```{r output-stuff}
exp_and_obs %>% 
  mutate(z_score = (obs_homoz - exp_homoz) / exp_SD) %>% 
  saveRDS(., file = "rds_outputs/exp_and_obs_homozygosity.rds", compress = "xz")
```




